resolve #305 abstract models

jose-padin opened the pull request:
Added support for abstract models. #305

jose-padin committed:
[#305] `ChordCounter` moved from `abstract` module to `generic` module.

Added some minor changes.

---

Resolves celery/django-celery-results#305

Fixes celery/django-celery-results#314

jose-padin committed:
[#305] Added abstract models.

`models` module replaced by a `models` package containing an `abstract`
module (for abstract models) and a `generic` module (for the default
models, previously in the `models` module)

---

Resolves celery/django-celery-results#305

Fixes celery/django-celery-results#314

jose-padin committed:
Issue 305: abstract models

* Fixed import bug

AllexVeldman reviewed (commented):
I'm not entirely clear on how this change is intended to be used.

When implementing a custom TaskResult model, how would you tell django_celery_results to use it instead of the included TaskResult? (I'm assuming this is the goal of this PR)

auvipy committed:
Update django_celery_results/models/__init__.py

Co-authored-by: Allex <a.veldman@chain-stock.com>

diegocastrum commented:
I think I've some idea to solve the issue you describe @AllexVeldman, I will work on that as soon as I can and share it with all of you to see what do you think about it, but for now I will be off for 3 weeks, so I will do it when I'm back.

Cheers!

hoefling commented:
@diegocastrum are you still on it or is the task up for grabs?

diegocastrum commented:
Hi @hoefling, sorry for not replying you earlier. I still working on it and I'm almost done with it. I already have a working scenario in one of my projects. So I expect to push the code very soon, I will try to do it this week. Thanks for asking!

diegocastrum committed:
[#305]: Improving abstract models implementation.

Added a `helpers` module into `models` containing the functions
`taskresult_model()` and `groupresult_model()`.

  * `taskresult_model()`: will try to find the custom model using a
    dotted path defined under the constant
`CELERY_RESULTS_TASKRESULT_MODEL` in the settings of the user's project
  * `groupresult_model()` will try to do the same using under the
    constant `CELERY_RESULTS_GROUPRESULT_MODEL`.

By default if these attributes are not found `django-celery-results`
will use the default models (`models.TaskResult` and
`models.GroupResult`).

Updated database backend in order to use custom models for `TaskResult
and `GroupResult` it they're present.

Instead to import explicitely the `TaskResult` and the `GroupResult`
(default models from `django-celery-results`) we make use of the model
helpers to load the right classes, the custom ones if they're present
otherwise we use the default ones.

Getting data from `task_kwargs` to extend the `task_properties` and be
able to store them into the database (using the custom models).

First of all we need a way to get data from `task_kwargs` (or somewhere
else) just before a `task_result` record is created, evaluate that data
and find the right values that will be used to fill the new fields
defined in the custom model.

So for this purpose we added a settings module to
`django-celery-results` which will hold default settings, the first
setting that will contain is a function in charge to get a callback from
the settings of the user project. This callback will be feeded by the
task `task_kwargs`, which will be intercepted in
`DatabaseBackend._get_extended_properties` just before the `task_kwargs`
are encoded by `encode_content()` method and send it to the
`store_result` method from the object manager of `TaskModel`
(Custom/Default one).

To end, we must to extend the arguments of the `store_result` method
from the `TaskResult` Manager adding `extra_fields` argument that will
make us able to send extra data to the custom model, when it's defined.

---
Resolves celery/django-celery-results#305

Fixes celery/django-celery-results#314

diegocastrum commented:
Hi @jose-padin, @AllexVeldman and @hoefling!

I already made a [commit](https://github.com/celery/django-celery-results/pull/314/commits/16be79384c774c6b2b2427aba93e0b993b98d417) improving the initial implementation and adding support to use behind the scenes the custom models defined in the project where `django-celery-results` is in use, to make use of this feature, the developer only needs define the next 3 constants in his settings module as follow:
  * Define `CELERY_RESULTS_TASKRESULT_MODEL` pointing to a custom model for `TaskResult`
  * Define `CELERY_RESULTS_GROUPRESULT_MODEL` pointing to a custom model for `GroupResult`
  * Define `CELERY_RESULTS_EXTEND_TASK_PROPS_CALLBACK` referencing a local function that takes kwargs as argument and returning a dict that will be used to store data into the new fields of the custom model referenced by `CELERY_RESULTS_TASKRESULT_MODEL`

This implementation is working right now in one of my projects and so far it seems to work fine. Probably we should add some checks, for example:
 * Some mechanic that forces the user to define the 3 constants instead of just one or two.
 * Maybe we could make use of the same callback to get parameters from the task `task_args`, but honestly I don't see the advantage of that right now.
 
 Let me know what you think & Cheers!

AllexVeldman commented:
@diegocastrum Looks like a viable implementation to me.

> Some mechanic that forces the user to define the 3 constants instead of just one or two.

Why is defining just 1 or 2 invalid? from what I can see, having only `CELERY_RESULTS_EXTEND_TASK_PROPS_CALLBACK` defined could cause weird behaviour.
If it makes sense to define `CELERY_RESULTS_TASKRESULT_MODEL` without `CELERY_RESULTS_EXTEND_TASK_PROPS_CALLBACK` I'm not entirely sure about, maybe if you only want to add/override functionality of the class without adding db columns.

AllexVeldman commented on django_celery_results/backends/database.py:
```diff
@@ -79,6 +79,10 @@ def _get_extended_properties(self, request, traceback):
                 # task protocol 1
                 task_kwargs = getattr(request, 'kwargs', None)
 
+            # TODO: We assuming that task protocol 1 could be always in use. :/
+            extended_props.update(
+                extend_task_props_callback(getattr(request, 'kwargs', None)))
```
Consider calling this from `_store_result()` so it can be used regardless of the extended result setting.

diegocastrum replied:
Nice point! It will be done in the next commit.

AllexVeldman commented on django_celery_results/settings.py:
```diff
@@ -0,0 +1,16 @@
+    if callable(callback):
+        return callback
+
+extend_task_props_callback = get_callback_function(
+    "CELERY_RESULTS_EXTEND_TASK_PROPS_CALLBACK"
+)
```
Consider adding a default callable returning an empty dict so this is an opt-in.
Currently this is mandatory if extended_result=True

diegocastrum replied:
Will be done in the next commit

AllexVeldman commented on django_celery_results/backends/database.py:
```diff
@@ -79,6 +79,10 @@ def _get_extended_properties(self, request, traceback):
 
+            # TODO: We assuming that task protocol 1 could be always in use. :/
+            extended_props.update(
+                extend_task_props_callback(getattr(request, 'kwargs', None)))
```
Consider passing `task_props` and `request` (in `_store_result()`) to the callback, this gives the developer more context to get whatever additional data it needs to store in the extended model.

diegocastrum replied:
That's clear, It will be done in the next commit.

AllexVeldman commented:
I'm not entirely sure how the docs are managed but I think this warrants a good description with some examples.

AllexVeldman commented on django_celery_results/settings.py:
```diff
@@ -0,0 +1,16 @@
+    if callable(callback):
+        return callback
+
+extend_task_props_callback = get_callback_function(
```
Consider adding some sanity checks on the return value of the callback.
For example that it complies to the `Mapping` protocol.

diegocastrum replied:
Well, `get_callback_function` is quite generic, and could be used in the future for another purposes, returning different types of data. So far I can see, the sanity checks could be in `_store_results()` where `extend_task_props_callback()` is called. What do you think?

AllexVeldman replied:
I'm personally not a fan of this generic function, I'd err of the side of caution and make the callback handling as clean and explicit as possible so any errors we need to raise have a clear source.
But this is more of a code-style thing so maybe one of the maintainers (@auvipy) can comment.

diegocastrum replied:
Good point @AllexVeldman, what do you think @auvipy?

diegocastrum replied:
@AllexVeldman & @auvipy I think I found a proper way to control the callback internally being able to check explicitly that return value.

I just created a new function called `get_task_props_extension()` into the settings module in charge to return an empty dict in case that the `CELERY_RESULTS_EXTEND_TASK_PROPS_CALLBACK` is undefined and otherwise will check that the return value complies with the Mapping protocol.

Let me know what you think!

AllexVeldman replied:
LGTM

diegocastrum committed:
[#305]: `extend_task_props_callback` relocated.

`extend_task_props_callback` moved from `_get_extended_properties`
to `_store_result`. Suggested by @AllesVeldman.

`extend_task_props_callback` will get the `request` object as first
parameter and a copy of `task_props` (avoiding potential manipulation of
the original `task_props`) as second paramenter.

---
Resolves celery/django-celery-results#305

Fixes celery/django-celery-results#314

auvipy reviewed (changes requested):
this will also need a rebase

diegocastrum committed:
[#305]: Added a default callable to `get_callback_function`

`get_callback_function()` gets a default callback as an arg
returning explicitely an empty dict.

`get_callback_function()` raises an `ImproperlyConfigured`
exception when the callback is not callable.

---
Resolves celery/django-celery-results#305

Fixes celery/django-celery-results#314

auvipy commented on django_celery_results/settings.py:
```diff
@@ -1,16 +1,19 @@
 extend_task_props_callback = get_callback_function(
-    "CELERY_RESULTS_EXTEND_TASK_PROPS_CALLBACK"
+    "CELERY_RESULTS_EXTEND_TASK_PROPS_CALLBACK", dict
 )
```
```suggestion
)

```

diegocastrum replied:
@auvipy This is already done. I couldn't apply your suggestion because the rebase but the change is already there.

diegocastrum committed:
Added newline to the end of

diegocastrum committed:
[#305] Added a sanity check to `task_props_extension`

Added `get_task_props_extension` to `settings` module which will raise
an `ImproperlyConfigured` when the task_props_extension doesn't complies
with the Mapping protocol.

`DatabaseBackend` will make use of `get_task_props_extension` to update
a potential custom model with the custom properties

---

Resolves celery/django-celery-results#305

Fixes celery/django-celery-results#314

diegocastrum committed:
Fixed a NoneType error when the callback is not defined in project settings.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

diegocastrum commented:
> this will also need a rebase

Done

AllexVeldman commented on django_celery_results/settings.py:
```diff
@@ -0,0 +1,36 @@
+    task_props_extension = extend_task_props_callback(request, task_props) or {}
+    if not isinstance(task_props_extension, Mapping):
+        raise ImproperlyConfigured(
+            "CELERY_RESULTS_EXTEND_TASK_PROPS_CALLBACK must return a Mapping "
+            "instance."
```
```suggestion
            "CELERY_RESULTS_EXTEND_TASK_PROPS_CALLBACK must return a Mapping."
```
since it's not an actual instance of the Mapping ABC, just something that implements it's interface.

diegocastrum committed:
Fixed a "wrong" description for the `ImproperlyConfigured` exception in `django_celery_results.settings`

Fixed a "wrong" description for the `ImproperlyConfigured` exception raised when `task_props_extension` doesn't complies with the Mapping protocol. At this point `task_props_extension` is just a dict that inherits from Mapping, not an explicit instance.

Thanks @AllexVeldman

Co-authored-by: Allex <a.veldman@chain-stock.com>

diegocastrum commented:
@auvipy or @AllexVeldman Could you give us an idea about when could be this merged?

AllexVeldman commented:
> Could you give us an idea about when could be this merged?

I'm not involved in merging/releasing, but I think there is still a bit of documentation missing, maybe extend https://github.com/celery/django-celery-results/blob/master/docs/getting_started.rst with an "Extending the results" part or creating a separate page for it.

diegocastrum commented:
> > Could you give us an idea about when could be this merged?
> 
> I'm not involved in merging/releasing, but I think there is still a bit of documentation missing, maybe extend https://github.com/celery/django-celery-results/blob/master/docs/getting_started.rst with an "Extending the results" part or creating a separate page for it.

You're definetly right. I'll try to do it asap

diegocastrum committed:
[#305] Added documentation about this feature.

A new page added to the documentation where we explain how to use this
feature.

--

issue: celery/django-celery-results#305

pull-request: celery/django-celery-results#314

diegocastrum committed:
[#305] Fixed some pre-commit failures

AllexVeldman commented on docs/extending_task_results.rst:
```diff
@@ -0,0 +1,43 @@
+Extending Task Results
+======================
+
+There are situations where you want to extend the Task Results with additional information that will make you able to retrieve information that was important at execution time of the task but not part of the task result itself. For example if you use :pypi:`django-celery-results` to track the task results from an tenant.
```
```suggestion
There are situations where you want to extend the Task Results with additional information that will make you able to retrieve information that was important at execution time of the task but not part of the task result itself. For example if you use :pypi:`django-celery-results` to track the task results from a tenant.
```

AllexVeldman commented on docs/extending_task_results.rst:
```diff
@@ -0,0 +1,43 @@
+
+            return {"tenant_id": task_kwargs.get("tenant_id", None)}
+
+#. To let :pypi:`django-celery-results` call this function internally you've to set the `CELERY_RESULTS_EXTEND_TASK_PROPS_CALLBACK` constant in your Django project's :file:`settings.py` with the function that you've just created.
```
```suggestion
#. To let :pypi:`django-celery-results` call this function, you'll have to set the `CELERY_RESULTS_EXTEND_TASK_PROPS_CALLBACK` constant in your Django project's :file:`settings.py`.
```

AllexVeldman commented on docs/extending_task_results.rst:
```diff
@@ -0,0 +1,43 @@
+        
+        CELERY_RESULTS_TASKRESULT_MODEL = 'myapp.TaskResult'
+
+#. Write a function in your Django project's :file:`settings.py` that will consume a `request` and `task_properties` as positional arguments and will return a dictionary with the additional information that you want to store in the your custom `TaskResult` model. The keys of this dictionary will be the fields of the custom model and the values the data you can retrieve from the `request` and/or ` ...
```
```suggestion
#. Write a function that will consume a `request` and `task_properties` as positional arguments and will return a dictionary with the additional information that you want to store in your custom `TaskResult` model. The keys of this dictionary will be the fields of the custom model and the values the data you can retrieve from the `request` and/or `task_properties`.
```

diegocastrum committed:
Update docs/extending_task_results.rst

Co-authored-by: Allex <a.veldman@chain-stock.com>

diegocastrum committed:
Update docs/extending_task_results.rst

Co-authored-by: Allex <a.veldman@chain-stock.com>

diegocastrum committed:
Update docs/extending_task_results.rst

Co-authored-by: Allex <a.veldman@chain-stock.com>

diegocastrum commented:
@auvipy Do you see anything else we've to do/work around in order to add this feature into the main branch?

diegocastrum commented:
@auvipy We don't hear anything from you since [Aug 16 2022](https://github.com/celery/django-celery-results/pull/314#pullrequestreview-1073459906). Please say something and give us some hope to get this merged.

auvipy commented:
This is a big change and need more work to get accepted

diegocastrum commented:
> This is a big change and need more work to get accepted

Thanks @auvipy, happy to hear some feedback. Could you give me some more specific clue about where should I put my focus? The last time I dive deep into this was a year ago and since then I used this in another project without having any issue

axsaucedo commented:
Is this still planned? From a user perspective it would be quite a useful addition to have this `AbstractTaskResult` class

auvipy commented:
> > This is a big change and need more work to get accepted
> 
> Thanks @auvipy, happy to hear some feedback. Could you give me some more specific clue about where should I put my focus? The last time I dive deep into this was a year ago and since then I used this in another project without having any issue

the remaining part is all about keeping backward compatibility for existing users

auvipy commented on django_celery_results/models/generic.py:
```diff
@@ -0,0 +1,78 @@
+        app_label = "django_celery_results"
+
+
+class ChordCounter(models.Model):
```
why we are not using any abstract model here?

diegocastrum replied:
There is no specific reason at the moment. Will be added in the next iteration/update

diegocastrum commented:
> > > This is a big change and need more work to get accepted
> > 
> > 
> > Thanks @auvipy, happy to hear some feedback. Could you give me some more specific clue about where should I put my focus? The last time I dive deep into this was a year ago and since then I used this in another project without having any issue
> 
> the remaining part is all about keeping backward compatibility for existing users

Thank you @auvipy glad to hear your input and @axsaucedo to show up the interest! I think I could work on this in about 3 or 4 weeks.

diegocastrum committed:
feat(models): add `AbstractChordCounter` and update `ChordCounter`

- Added new abstract model `AbstractChordCounter` in `abstract.py` for Chord synchronization, including fields for `group_id`, `sub_tasks`, and `count`.
- Updated `ChordCounter` model in `generic.py` to inherit from `AbstractChordCounter`.
- Moved `group_result` method from `ChordCounter` to `AbstractChordCounter`.
- Added helper function `chordcounter_model` in `helpers.py` to return the active `ChordCounter` model.
- Updated import statements in `database.py` to use `chordcounter_model` helper function.
- Added `ChordCounterModel` attribute to `DatabaseBackend` class, and updated `create` and `get` methods to use `ChordCounterModel` instead of `ChordCounter`.

Relates to: #305

diegocastrum committed:
fix: refactor helper functions to avoid circular dependencies

diegocastrum committed:
fix: undefined name 'ChordCounter' and minor fixes

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

diegocastrum commented:
> this will also need a rebase

done

diegocastrum committed:
fix: 'TypeError' introduced in previous commit

diegocastrum committed:
fix: include 'extra_fields' conditionally

diegocastrum committed:
fix: 'get_task_props_extensions()' missing 1 required argument

diegocastrum committed:
fix: TypeError introducedn on prev commit on 'AbstractChordCounter'

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

diegocastrum committed:
fix: ImportError introduced in previous commit in 'abstract.py'

diegocastrum committed:
style: Reformat import statements for better readability in 'database.py'

diegocastrum committed:
fix: Update configuration for isort and black to enforce line length and multi-line output

This commit addresses an issue where the pre-commit hook merges import formatted with isort's multiline output mode 3 back into a single line, resulting in a flake8 E508 violation in CI due to lines exceeding 79 characters

diegocastrum commented:
Friendly reminder: @auvipy Could we move forward with this `merge|pull` request `and|or` `check|review` it to see if something is missing, in that case I would like to get some feedback

diegocastrum commented:
Thank you @auvipy! Do you have an idea/plan about how are `you(?:/we)?` going to proceed from now on regarding this feature?

browniebroke reviewed (commented):
I think this would be a nice addition to the library 👍🏻 

I left a small suggestion to make use of swappable models which is perfect for this use case.

diegocastrum commented:
> I think this would be a nice addition to the library 👍🏻
> 
> I left a small suggestion to make use of swappable models which is perfect for this use case.

Thank you so much @browniebroke for your feedback! And sorry for my late reaction, I missed your message. I will check your suggestions asap

diegocastrum committed:
Make generic models swappable

Suggestions from code review

Co-authored-by: Bruno Alla <browniebroke@users.noreply.github.com>

diegocastrum committed:
refactor: Rename model retrieval functions for consistency to align with django auth's app.

Ref: https://github.com/celery/django-celery-results/pull/314#discussion_r2395047841

diegocastrum commented on django_celery_results/models/helpers.py:
```diff
@@ -0,0 +1,72 @@
+from .generic import ChordCounter, GroupResult, TaskResult
+
+
+def taskresult_model():
```
@browniebroke Thanks! I just added support to swap models like Django auth's app does

> I would follow Django auth's app establish pattern from `get_user_model` here and call these `get_task_result_model()`, `get_chord_counter_model()` and `get_group_result_model()`, but that's perhaps quite subjective, so I'll let the maintainers decide.

@browniebroke Actually I also like the idea, let's see what @maintainers think about that

diegocastrum commented:
> [I will handle it myself as part of big refactor](https://github.com/celery/django-celery-results/pull/314#issuecomment-2731754146)

@auvipy Just checking if you have any insights on this

diegocastrum committed:
feat: Add ALL_STATES and TASK_STATE_CHOICES to __all__ export; update docstring argument format

diegocastrum committed:
feat: Add documentation for abstract, generic, helpers, settings, urls, and views modules

diegocastrum commented:
Hi @auvipy, I’ve updated this PR branch again and tried to focus specifically on the concerns you raised earlier.

Main points addressed:

- Rebased on current `upstream/main`.
- Added abstract models for the result models, including `AbstractTaskResult`, `AbstractGroupResult`, and `AbstractChordCounter`.
- Kept the existing concrete models available through the old import path:
  `from django_celery_results.models import TaskResult, GroupResult, ChordCounter`
- Preserved backwards compatibility for existing `django_celery_results.models` imports, including the previously exported constants.
- No migration/schema change is introduced for existing users.
- Added swappable model support for `TaskResult`, `GroupResult`, and `ChordCounter`, following the direction suggested in review.
- Added docs for extending task results and API reference coverage for the new modules.
- Docker tox matrix passes locally.

This PR has been open for quite a long time, and there has been repeated user interest in the feature. I’d like to help move it to a clear outcome, whether that means final review or splitting it up.

Given your earlier comment that the remaining work was mainly about backwards compatibility, could you let me know if this now addresses that concern?

If the PR is still too large to review/merge as-is, I can split it into smaller PRs, for example:

1. Extract abstract models while preserving existing public imports.
2. Add swappable model support.
3. Add the task property extension callback.
4. Add docs/tests.

I’d appreciate guidance on which path would make this easiest for maintainers to review.

diegocastrum committed:
fix: store extended task result fields

diegocastrum committed:
fix: validate custom task result fields

diegocastrum committed:
docs: clarify custom task result fields

diegocastrum committed:
refactor: deduplicate result model helpers
